.github/workflows/main.yml

```diff
@@ -140,33 +140,172 @@ jobs:
+        run: cp $GITHUB_WORKSPACE/bazel-bin/oppia.apk /home/runner/work/oppia-android/oppia-android/
       - uses: actions/upload-artifact@v2
         with:
           name: oppia-bazel-apk
           path: /home/runner/work/oppia-android/oppia-android/oppia.apk
+
+  bazel_compute_affected_targets:
+    name: Compute affected tests
+    runs-on: ubuntu-18.04
+    outputs:
+      matrix: ${{ steps.compute-test-matrix-from-affected.outputs.matrix || steps.compute-test-matrix-from-all.outputs.matrix }}
+    steps:
       - uses: actions/checkout@v2
-      
+        with:
+          fetch-depth: 0
+      - name: Clone Oppia Bazel
+        run: git clone https://github.com/oppia/bazel.git $HOME/oppia-bazel
+      - name: Unzip Bazel binary
+        run: |
+          cd $HOME/oppia-bazel
+          unzip bazel-build.zip
+          cd $GITHUB_WORKSPACE
+          chmod a+x $HOME/oppia-bazel/bazel
+      - name: Compute test matrix based on affected targets
+        id: compute-test-matrix-from-affected
+        if: "!contains(github.event.pull_request.title, '[RunAllTests]')"
+        # https://unix.stackexchange.com/a/338124 for reference on creating a JSON-friendly
+        # comma-separated list of test targets for the matrix.
+        run: |
+          TEST_TARGET_LIST=$(bash ./scripts/compute_affected_tests.sh $HOME/oppia-bazel/bazel | sed 's/^\|$/"/g' | paste -sd, -)
+          echo "Affected tests (note that this might be all tests if on the develop branch): $TEST_TARGET_LIST"
+          echo "::set-output name=matrix::{\"test-target\":[$TEST_TARGET_LIST]}"
+      - name: Compute test matrix based on all tests
+        id: compute-test-matrix-from-all
+        if: "contains(github.event.pull_request.title, '[RunAllTests]')"
+        run: |
+          TEST_TARGET_LIST=$($HOME/oppia-bazel/bazel query "kind(test, //...)" | sed 's/^\|$/"/g' | paste -sd, -)
+          echo "Affected tests (note that this might be all tests if on the develop branch): $TEST_TARGET_LIST"
+          echo "::set-output name=matrix::{\"test-target\":[$TEST_TARGET_LIST]}"
+
+  bazel_run_test:
+    name: Run Bazel Test
+    needs: bazel_compute_affected_targets
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix: ${{fromJson(needs.bazel_compute_affected_targets.outputs.matrix)}}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Clone Oppia Bazel
+        run: git clone https://github.com/oppia/bazel.git $HOME/oppia-bazel
+      - name: Set up JDK 9
+        uses: actions/setup-java@v1
+        with:
+          java-version: 9
+      - name: Extract Android tools
+        run: |
+          mkdir -p $GITHUB_WORKSPACE/tmp/android_tools
+          cd $HOME/oppia-bazel
+          unzip bazel-tools.zip
+          tar -xf $HOME/oppia-bazel/android_tools.tar.gz -C $GITHUB_WORKSPACE/tmp/android_tools
+      # See explanation in bazel_build_app for how this is installed.
+      - name: Install git-secret (non-fork only)
```

Comment:
This step is being used in multiple places, can we maybe define the step (or the block of code ) once and re-use it below?

Reply:
See my reply to @anandwana001's comment above--it's not straightforward to reuse stuff in actions (unless it can be put in a script, and we need to be really careful with the secret so that we don't accidentally print it in plaintext in the Bash history file or as part of the process command line visible via ``ps``.